File: mc302/Lab04/br/unicamp/ic/mc302/termostato/Termostato.java
package br.unicamp.ic.mc302.termostato;

public class Termostato extends Termometro{
	private Termometro termometro;
	private Chave chave;
	private float tempRequerida;
	
	public Termostato(){
		termometro = new Termometro();
		chave = new Chave(); 
	}
	
	public void setTempRequerida(float temp){
		tempRequerida = temp; 
	}
	
	public void fazerMonitoramento(){
		if (termometro.getTemperatura()>tempRequerida)
			chave.desligar();
		else
			chave.ligar(); 
	}
	
	public boolean chaveLigada() {
		return chave.ligada();
	}
}

File: mc302/Lab04/br/unicamp/ic/mc302/termostato/Termostato.cpp
#include <iostream>
using namespace std;

class Termometro {
	private:
		int temperatura;
	public:
		int getTemperatura() {
			return temperatura;
		}
		int setTemperatura(int nTemp) {
			temperatura = nTemp;
		}
};

class Chave {
	private:
		int estadoChave; // 0 - desligada; 1 - ligada
	public:
		void ligar() {
			this -> estadoChave = 1;
		}
		void desligar() {
			this -> estadoChave = 0;
		}
		int getEstadoChave() {
			return estadoChave;
		}
};

class Termostato: public Termometro, public Chave {
	private:
		int tempRequerida;
	public:
		Termostato(int nTemp) {
			tempRequerida = nTemp;
		}
		void setTempRequerida(int nTemp) {
			if(nTemp > 0)
				tempRequerida = nTemp;
		}
		void fazerMonitoramento() {
			if(getTemperatura() < tempRequerida) {
				ligar();
			} else {
				desligar();
			}
		}
};

int main() {
	Termostato *t = new Termostato(25);
	t->desligar();
	t->setTemperatura(20);

	cout << "estado inicial: " << t->getEstadoChave() << endl;
	t->fazerMonitoramento();
	cout << "estado final: " << t->getEstadoChave() << endl;

	return 0;
}
